### PR TITLE
Rename Tomcat metrics, fix types and add help text

### PR DESCRIPTION
**Pull request by elecnix**
Rename most of the interesting Tomcat JMX metrics to follow naming convention of Prometheus, specifying type and help text.

This has been tested with Tomcat 8.

### CODE REVIEW

File: example_configs/tomcat.yml [hunk 1]
```diff
@@ -1,33 +1,680 @@
 ---   
+# Example jmx_exporter configuration for use with Tomcat 8.
+
 lowercaseOutputLabelNames: true
 lowercaseOutputName: true
-rules:
-- pattern: 'Catalina<type=GlobalRequestProcessor, name=\"(\w+-\w+)-(\d+)\"><>(\w+):'
-  name: tomcat_$3_total
+
+whitelistObjectNames: ["Tomcat:*", "java.lang:*"]
```
Comment (brian-brazil) [inline example_configs/tomcat.yml]: Don't include java.lang, they're already taken care of.

File: example_configs/tomcat.yml [hunk 2]
```diff
@@ -1,33 +1,680 @@
...
+
+# ------------------------ GlobalRequestProcessor -----------------------
+
+# maxTime
+- pattern: '^Tomcat<type=GlobalRequestProcessor, name=\"(.*)\"><>maxTime:'
+  name: tomcat_global_processingtime_milliseconds_max
+  help: 'Tomcat global maximum request processing time seen, in milliseconds.'
```
Comment (brian-brazil) [inline example_configs/tomcat.yml]: There was a recent change to allow multiplication, so you can convert this to seconds.

I doubt that max is the mertric you want though.

File: example_configs/tomcat.yml (hunk 1 above)
Comment (elecnix) [inline example_configs/tomcat.yml]: How are they taken care of?

File: example_configs/tomcat.yml (hunk 2 above)
Comment (elecnix) [inline example_configs/tomcat.yml]: Thanks for the reminder. I'll do the multiplication.

Do you mean the max is not very useful?

File: example_configs/tomcat.yml (hunk 1 above)
Comment (brian-brazil) [inline example_configs/tomcat.yml]: See https://github.com/prometheus/client_java/tree/master/simpleclient_hotspot

**Comment by elecnix** [comment]
Brian, the list is not exhaustive. Many were omitted, and others were just enumerated in comments. I also tried using regular expressions to catch many in one rule, but trying to balance terseness, useful help text, and maintainability. How do you propose to collapse it down?

**Comment by brian-brazil** [comment]
Help text isn't that important, 10ish general rules should be sufficient to cover most servers.

File: example_configs/tomcat.yml (hunk 1 above)
Comment (elecnix) [inline example_configs/tomcat.yml]: Ok, I'll drop them all.

Some are important to me, but I understand it doesn't belong here.

Comment (brian-brazil) [inline example_configs/tomcat.yml]: If there's some missing I'll take a PR over in client_java.

File: example_configs/tomcat.yml [hunk 3]
```diff
@@ -1,33 +1,680 @@
...
+- pattern: '^Tomcat<type=Manager, host=(.*), context=(.*)><>sessionMaxAliveTime:'
+  type: GAUGE
+  name: tomcat_sessions_active_length_seconds_max
+  labels:
+    context: "$2"
+
+# Number of active sessions at this moment (Tomcat<type=Manager, host=localhost, context=/><>activeSessions)
+- pattern: '^Tomcat<type=Manager, host=(.*), context=(.*)><>activeSessions:'
+  type: GAUGE
+  name: tomcat_sessions_active_total
```
Comment (sdurrheimer) [inline example_configs/tomcat.yml]: `_total` suffix should be reserved to counters.

File: example_configs/tomcat.yml [hunk 4]
```diff
@@ -1,33 +1,680 @@
...
+- pattern: '^Tomcat<type=Manager, host=(.*), context=(.*)><>(\w*)Sessions:'
+  type: COUNTER
+  name: tomcat_sessions_$3
+  labels:
+    context: "$2"
+
+# Total number of sessions created by this manager (Tomcat<type=Manager, host=localhost, context=/><>sessionCounter)
+- pattern: '^Tomcat<type=Manager, host=(.*), context=(.*)><>sessionCounter:'
+  type: COUNTER
+  name: tomcat_sessions_created
```
Comment (sdurrheimer) [inline example_configs/tomcat.yml]: `_total`

File: example_configs/tomcat.yml [hunk 5]
```diff
@@ -1,33 +1,680 @@
...
+  labels:
+    context: "$2"
+
+# Number of duplicated session ids generated (Tomcat<type=Manager, host=localhost, context=/><>duplicates)
+- pattern: '^Tomcat<type=Manager, host=(.*), context=(.*)><>duplicates:'
+  type: COUNTER
+  name: tomcat_sessions_duplicated
```
Comment (sdurrheimer) [inline example_configs/tomcat.yml]: same `_total`

File: example_configs/tomcat.yml [hunk 6]
```diff
@@ -1,33 +1,680 @@
...
+    context: $1
+
+# Amount of ms that the container will wait for servlets to unload (Tomcat<j2eeType=WebModule, name=//localhost/, J2EEApplication=none, J2EEServer=none><>unloadDelay)
+- pattern: '^Tomcat<j2eeType=WebModule, name=//localhost(.*), J2EEApplication=none, J2EEServer=none><>unloadDelay:'
+  name: tomcat_webmodule_unloadDelay_milliseconds
+  type: GAUGE
+  labels:
+    context: $1
+
+# ------------------------ java.lang -----------------------
```
Comment (sdurrheimer) [inline example_configs/tomcat.yml]: I'm not sure this should be hidden  at the bottom of the tomcat example.

File: example_configs/tomcat.yml [hunk 7]
```diff
@@ -1,33 +1,680 @@
...
+
+- pattern: '^java.lang<type=Threading><>TotalStartedThreadCount:'
+  name: java_threading_thread_started
+  help: 'The total number of threads created and also started since the Java virtual machine started.'
+  type: COUNTER
+
+# ------------------------ WILDCARD -----------------------
+
+# This ensures we export all metrics, even if one of the above regexes fail, or if new metrics are added.
+- pattern: ".*"
```
Comment (sdurrheimer) [inline example_configs/tomcat.yml]: Same here, i'm 👎 having this here.

**Comment by hrecho** [comment]
hi guys, i am sorry if this is not the right place to ask
i am getting javax.management.MalformedObjectNameException: Unterminated key property part
while trying to ...
new ObjectName("Tomcat:type=ThreadPool,name=\"http-nio-8080\"/currentThreadsBusy");
do you have any hint for me?
thanks in advance

**Comment by hrecho** [comment]
quotas are escaped ofcourse
`new ObjectName("Tomcat:type=ThreadPool,name=\"http-nio-8080\"/currentThreadsBusy");`

File: example_configs/tomcat.yml [hunk 8]
```diff
@@ -1,33 +1,409 @@
...
+# ------------------------ StringCache -----------------------
+
+- pattern: '^Tomcat<type=StringCache><>accessCount:'
+  name: tomcat_stringcache_access_total
+  help: 'Tomcat: The number of times the string cache was accessed.'
+  type: COUNTER
+- pattern: '^Tomcat<type=StringCache><>cacheSize:'
+  name: tomcat_stringcache_size
+  valueFactor: 0.001
+  help: 'Tomcat: The number of entries in the string cache.'
```
Comment (brian-brazil) [inline example_configs/tomcat.yml]: What's the unit here? It's odd that a count would be divided by 1000

File: example_configs/tomcat.yml [hunk 9]
```diff
@@ -1,33 +1,409 @@
...
+  valueFactor: 0.001
+  type: COUNTER
+  labels:
+    context: "$1"
+    servlet: "$2"
+
+# Maximum processing time of a request (Tomcat<j2eeType=Servlet, WebModule=//localhost/, name=default, J2EEApplication=none, J2EEServer=none><>maxTime)
+# Minimum processing time of a request (Tomcat<j2eeType=Servlet, WebModule=//localhost/, name=default, J2EEApplication=none, J2EEServer=none><>minTime)
+- pattern: '^Tomcat<j2eeType=Servlet, WebModule=//localhost(.*), name=(.*), J2EEApplication=none, J2EEServer=none><>(max|min)Time:'
+  name: tomcat_servlet_processing_$3_seconds
```
Comment (brian-brazil) [inline example_configs/tomcat.yml]: Be consistent, you've the min/max at the end above

File: example_configs/tomcat.yml [hunk 10]
```diff
@@ -1,33 +1,409 @@
...
+# These already have a help text, so we let it pass through.
+
+# Cumulative error count of all servlets in this context (Tomcat<j2eeType=WebModule, name=//localhost/, J2EEApplication=none, J2EEServer=none><>errorCount)
+# Cumulative request count of all servlets in this context (Tomcat<j2eeType=WebModule, name=//localhost/, J2EEApplication=none, J2EEServer=none><>requestCount)
+- pattern: '^Tomcat<j2eeType=WebModule, name=//localhost(.*), J2EEApplication=none, J2EEServer=none><>requestCount:'
+  name: tomcat_webmodule_requests_total
+  type: COUNTER
+  labels:
+    context: $1
+- pattern: '^Tomcat<j2eeType=WebModule, name=//localhost(.*), J2EEApplication=none, J2EEServer=none><>errorCount:'
```
Comment (brian-brazil) [inline example_configs/tomcat.yml]: This way of extracting context is different to the ones above. Will it always have the same content and meaning?

File: example_configs/tomcat.yml [hunk 11]
```diff
@@ -1,33 +1,409 @@
...
+  name: tomcat_webmodule_requests_errors_total
+  type: COUNTER
+  labels:
+    context: $1
+
+# Maximum execution time of all servlets in this context (Tomcat<j2eeType=WebModule, name=//localhost/, J2EEApplication=none, J2EEServer=none><>maxTime)
+# Minimum execution time of all servlets in this context (Tomcat<j2eeType=WebModule, name=//localhost/, J2EEApplication=none, J2EEServer=none><>minTime)
+- pattern: '^Tomcat<j2eeType=WebModule, name=//localhost(.*), J2EEApplication=none, J2EEServer=none><>(min|max)Time:'
+  name: tomcat_webmodule_execution__$2
```
Comment (brian-brazil) [inline example_configs/tomcat.yml]: Extra _

File: example_configs/tomcat.yml [hunk 12]
```diff
@@ -1,33 +1,409 @@
 ---   
+# Example jmx_exporter configuration for use with Tomcat 8.
+
 lowercaseOutputLabelNames: true
 lowercaseOutputName: true
-rules:
```
Comment (Teriand) [inline example_configs/tomcat.yml]: rules need comeback, or  error:
Caused by: while parsing a block mapping
 in 'reader', line 4, column 1:
    lowercaseOutputLabelNames: true
    ^
expected <block end>, but found BlockEntry
 in 'reader', line 12, column 1:
    - pattern: '^Tomcat<type=GlobalR ...

**Comment by brian-brazil** [comment]
This has gone stale, so closing for now.

**Pull request closed without merging**